### PR TITLE
Reference Implementation of ROIPooling op

### PR DESCRIPTION
**Pull request by ggalieroc**
Overview:
- Update specification
- Develop ROIPooling reference implementation
- Add reference implementation to int_executable
- Add unit tests
- Create single layer test class
- Instantiate single layer tests for CPU and Myriad plugins

ticket: CVS-37530

### CODE REVIEW

**Comment by iefode** [comment]
Please, take Single Layer test from [PR](https://github.com/openvinotoolkit/openvino/pull/2742), refactor this if it will be necessary and instantiate for CPU plugin

File: docs/ops/detection/ROIPooling_1.md [hunk 1]
```diff
@@ -44,13 +44,17 @@
 
 **Inputs**:
 
-*   **1**: 4D input tensor of shape `[1, C, H, W]` with feature maps. Required.
+*   **1**: 4D input tensor of shape `[1, C, H, W]` with feature maps of type *T*. Required.
 
-*   **2**: 2D input tensor of shape `[NUM_ROIS, 5]` describing box consisting of 5 element tuples: `[batch_id, x_1, y_1, x_2, y_2]`. Required.
+*   **2**: 2D input tensor of shape `[NUM_ROIS, 5]` describing box consisting of 5 element tuples: `[batch_id, x_1, y_1, x_2, y_2]` in relative coordinates of type *T*. The box height and width are calculated the following way: `roi_width = max(spatial_scale * (x_2 - x_1), 1.0)`, `roi_height = max(spatial_scale * (y_2 - y_1), 1.0)`, so the malformed boxes are expressed as a box of size `1 x 1`. Required.
```
Comment (lazarevevgeny) [inline docs/ops/detection/ROIPooling_1.md:49]: I am not 100% sure but when the mode is "bilinear" the coordinates are absolute.

@dmitry-gorokhov , @vladimir-paramuzov , could you comment on that?

File: docs/ops/detection/ROIPooling_1.md [hunk 2]
```diff
@@ -44,13 +44,17 @@
 
 **Inputs**:
 
-*   **1**: 4D input tensor of shape `[1, C, H, W]` with feature maps. Required.
+*   **1**: 4D input tensor of shape `[1, C, H, W]` with feature maps of type *T*. Required.
```
Comment (lazarevevgeny) [inline docs/ops/detection/ROIPooling_1.md:47]: The batch size of this input tensor may be arbitrary number `N` corresponding to number of input images.

The second tensor input with tuples `[batch_id, x_1, y_1, x_2, y_2]` values for elements `batch_id` must be in the range `[0, batch_id - 1]`

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: docs/ops/detection/ROIPooling_1.md [hunk 3]
```diff
@@ -44,9 +44,15 @@
 
 **Inputs**:
 
-*   **1**: 4D input tensor of shape `[1, C, H, W]` with feature maps of type *T*. Required.
+*   **1**: 4D input tensor of shape `[N, C, H, W]` with feature maps of type *T*. Required.
+
+*   **2**: 2D input tensor of shape `[NUM_ROIS, 5]` describing region of interest box consisting of 5 element tuples: `[batch_id, x_1, y_1, x_2, y_2]`. Required.
+Batch indices must be in the range of [0, N-1].
```
Comment (lazarevevgeny) [inline docs/ops/detection/ROIPooling_1.md:50]: surround [0, N-1] with "`" characters

File: ngraph/core/include/ngraph/op/roi_pooling.hpp [hunk 4]
```diff
@@ -31,42 +31,47 @@ namespace ngraph
...
                 /// \param coords         Coordinates of bounding boxes
-                /// \param output_size    Height/Width of ROI output features
+                /// \param pooled_h       Height of the ROI output features
+                /// \param pooled_w       Width of the ROI output features
                 /// \param spatial_scale  Ratio of input feature map over input image size
                 /// \param method         Method of pooling - Max or Bilinear
                 ROIPooling(const Output<Node>& input,
                            const Output<Node>& coords,
-                           const Shape& output_size,
+                           const int pooled_h,
```
Comment (lazarevevgeny) [inline ngraph/core/include/ngraph/op/roi_pooling.hpp:49]: Changing operation semantic requires introduction of a new operation. We cannot modify the API of existing operation! Please, revert these changes.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: ngraph/core/include/ngraph/op/roi_pooling.hpp (hunk 4 above)
Comment (ggalieroc) [reply ngraph/core/include/ngraph/op/roi_pooling.hpp]: Done

File: docs/ops/detection/ROIPooling_1.md (hunk 3 above)
Comment (ggalieroc) [reply docs/ops/detection/ROIPooling_1.md]: Done

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_pooling.hpp [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+            void roi_pooling(const T* feature_maps,
+                             const T* rois,
+                             T* output,
+                             const Shape& feature_maps_shape,
+                             const Shape& rois_shape,
+                             const Shape& output_shape,
+                             const float spatial_scale,
+                             const ROIPoolingMethod& pooling_method)
+            {
+                const int channels = feature_maps_shape[1];
```
Comment (pelszkow) [inline ngraph/core/reference/include/ngraph/runtime/reference/roi_pooling.hpp:38]: 👍 
Can we use `const` elsewhere.

File: inference-engine/tests/functional/plugin/shared/include/single_layer_tests/roi_pooling.hpp [hunk 6]
```diff
@@ -0,0 +1,47 @@
+// Copyright (C) 2020 Intel Corporation
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/shared/include/single_layer_tests/roi_pooling.hpp:1]: Where is instances?

**Review by iefode — APPROVED** [review]

**Comment by iefode** [comment]
And one more comment, could you please add instantiation for Myriad plugin?

**Comment by ggalieroc** [comment]
After discussion I decided to remove enum class ROIPoolingMethod to specify ROI pool method and leave it as string. Last commit address refactor single layer test class to align with this

File: docs/ops/detection/ROIPooling_1.md [hunk 7]
```diff
@@ -44,13 +50,22 @@
 
 **Inputs**:
 
-*   **1**: 4D input tensor of shape `[1, C, H, W]` with feature maps. Required.
+*   **1**: 4D input tensor of shape `[N, C, H, W]` with feature maps of type *T*. Required.
+
+*   **2**: 2D input tensor of shape `[NUM_ROIS, 5]` describing region of interest box consisting of 5 element tuples of type *T*: `[batch_id, x_1, y_1, x_2, y_2]`. Required.
```
Comment (lazarevevgeny) [inline docs/ops/detection/ROIPooling_1.md:55]: These changes are rendered very poorly. Let's move detailed description of the input format to the `detailed description` section

File: ngraph/test/type_prop/roi_pooling.cpp [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+TEST(type_prop, roi_pooling_invalid_pooled_size)
+{
+    const auto feat_maps = make_shared<op::Parameter>(element::f32, Shape{3, 2, 6, 6});
+    const auto rois = make_shared<op::Parameter>(element::f16, Shape{3, 5});
+    // ROIPooling pooled_h and pooled_w must be non-negative integers
+    ASSERT_THROW(make_shared<op::v0::ROIPooling>(feat_maps, rois, Shape{1, 0}, 0.625f, "max"),
+                 ngraph::NodeValidationFailure);
+}
```
Comment (lazarevevgeny) [inline ngraph/test/type_prop/roi_pooling.cpp:103]: Please, add various tests when one (or both) inputs have dynamic dimensions.
Also, need a test when inputs have dynamic rank.

File: ngraph/core/src/op/roi_pooling.cpp [hunk 9]
```diff
@@ -36,32 +36,81 @@ op::ROIPooling::ROIPooling(const Output<Node>& input,
...
+        "The spatial scale attribute should be a positive floating point number. Got: ",
+        m_spatial_scale);
+
+    NODE_VALIDATION_CHECK(
+        this,
+        m_method == "max" || m_method == "bilinear",
+        "Pooling method attribute should be either \'max\' or \'bilinear\'. Got: ",
+        m_method);
+
     if (get_input_partial_shape(0).is_static() && get_input_partial_shape(1).is_static())
```
Comment (lazarevevgeny) [inline ngraph/core/src/op/roi_pooling.cpp:69]: This check is too rough. We can calculate some of the output dimensions if `get_input_partial_shape(0)` or `get_input_partial_shape(1)` has **some** dynamic dimensions. Please, make the shape inference function more specific and try to calculate as much output dimensions as possible. And cover these changes with several shape inference unit tests.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]
The main comment is that the shape inference function must be more specific and try to calculate output shape even if some of the input dimensions are dynamic.

File: inference-engine/tests/functional/plugin/shared/src/single_layer_tests/roi_pooling.cpp [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        InferenceEngine::Precision netPrecision;
+        float spatial_scale;
+
+        std::tie(inputShape, coordsShape, poolShape, spatial_scale, pool_method, netPrecision, targetDevice) = this->GetParam();
+
+        auto ngPrc = FuncTestUtils::PrecisionUtils::convertIE2nGraphPrc(netPrecision);
+        auto params = ngraph::builder::makeParams(ngPrc, {inputShape, coordsShape});
+        auto paramOuts = ngraph::helpers::convert2OutputVector(
+                ngraph::helpers::castOps2Nodes<ngraph::op::Parameter>(params));
+        std::shared_ptr<ngraph::Node> roi_pooling = ngraph::builder::makeROIPooling(paramOuts[0],
```
[the remark this answers is not in the document]
Comment (ggalieroc) [reply inference-engine/tests/functional/plugin/shared/src/single_layer_tests/roi_pooling.cpp:92]: This function is needed to convert pooling method from enum to string and in case of invalid method throws runtime exception. I would leave it like this for now. There is a ticket to create a new version of ROIPooling supporting strong type for pooling method, this can be removed once that version is implemented.

File: inference-engine/src/legacy_api/src/ie_cnn_layer_builder_ngraph.cpp [hunk 11]
```diff
@@ -42,6 +42,7 @@
 #include <ngraph/ngraph.hpp>
 #include <ngraph/variant.hpp>
 #include <ngraph/opsets/opset5.hpp>
+#include <ngraph/enum_names.hpp>
```
[the remark this answers is not in the document]
Comment (ggalieroc) [reply inference-engine/src/legacy_api/src/ie_cnn_layer_builder_ngraph.cpp]: Removed.

File: ngraph/test/type_prop/roi_pooling.cpp (hunk 8 above)
Comment (ggalieroc) [reply ngraph/test/type_prop/roi_pooling.cpp:136]: Done

File: ngraph/core/src/op/roi_pooling.cpp (hunk 9 above)
Comment (ggalieroc) [reply ngraph/core/src/op/roi_pooling.cpp]: Done

**Review by lazarevevgeny — APPROVED** [review]
The spec changes and partial inference function LGTM.

File: ngraph/core/include/ngraph/op/roi_pooling.hpp [hunk 12]
```diff
@@ -32,7 +32,7 @@ namespace ngraph
                 ROIPooling() = default;
                 /// \brief Constructs a ROIPooling operation
                 ///
```
Comment (ilyachur) [inline ngraph/core/include/ngraph/op/roi_pooling.hpp:34]: Can we introduce `ROIPoolingTypes` and create a constructor with this enum?

File: inference-engine/tests/ngraph_functions/src/roi_pooling.cpp [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+#include <vector>
+#include <memory>
+
+#include "ngraph_functions/builders.hpp"
+
+namespace ngraph {
+namespace builder {
+
+std::shared_ptr<Node> makeROIPooling(const Output<Node>& input,
```
Comment (ilyachur) [inline inference-engine/tests/ngraph_functions/src/roi_pooling.cpp:14]: If we introduce new constructor it should allow to avoid this builder.

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_pooling.hpp [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+                                    h_end = std::min(std::max(h_end + roi_h_start, 0), height);
+                                    w_end = std::min(std::max(w_end + roi_w_start, 0), width);
+
+                                    const size_t pool_index =
+                                        roi_num * channels * pooled_h * pooled_w +
+                                        c * pooled_h * pooled_w + ph * pooled_w + pw;
+
+                                    // Define an empty pooling region to be zero
+                                    bool is_empty = (h_end <= h_start) || (w_end <= w_start);
+                                    if (is_empty)
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/runtime/reference/roi_pooling.hpp:114]:
```suggestion
                                    if (h_end <= h_start || w_end <= w_start)
```

File: inference-engine/tests/ngraph_functions/src/roi_pooling.cpp (hunk 13 above)
Comment (ggalieroc) [reply inference-engine/tests/ngraph_functions/src/roi_pooling.cpp:14]: As agreed in internal discussion, supporting strong type for ROIPooling method will be addressed in a different ticket: 43566

File: ngraph/core/include/ngraph/op/roi_pooling.hpp (hunk 12 above)
Comment (ggalieroc) [reply ngraph/core/include/ngraph/op/roi_pooling.hpp:34]: As agreed in internal discussion, supporting strong type for ROIPooling method will be addressed in a different ticket: 43566

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_pooling.hpp (hunk 14 above)
Comment (pelszkow) [reply ngraph/core/reference/include/ngraph/runtime/reference/roi_pooling.hpp]: @ilyachur , Why we can't use temporary boolean variables?

Comment (ilyachur) [reply ngraph/core/reference/include/ngraph/runtime/reference/roi_pooling.hpp]: We use this variable only in one place.
And why do we need to add a new variable if we can do it in-place?

Comment (jdanieck) [reply ngraph/core/reference/include/ngraph/runtime/reference/roi_pooling.hpp]: To give a meaningful name to this OR condition. I personally like it.

Comment (pelszkow) [reply ngraph/core/reference/include/ngraph/runtime/reference/roi_pooling.hpp]: No difference in asm: https://godbolt.org/z/vG7rb3

**Review by ilyachur — APPROVED** [review]
